src/transaction/atomic_replication_helper.hpp

```diff
@@ -158,47 +158,87 @@ namespace cublog
+
+	  struct page_ptr_info
+	  {
+	    page_ptr_info () = default;
+
+	    page_ptr_info (const page_ptr_info &) = delete;
+	    page_ptr_info (page_ptr_info &&) = default;
+
+	    page_ptr_info &operator= (const page_ptr_info &) = delete;
+	    page_ptr_info &operator= (page_ptr_info &&) = delete;
+
+	    ~page_ptr_info ();
+
+	    VPID m_vpid = VPID_INITIALIZER;
+	    LOG_RCVINDEX m_rcv_index = RV_NOT_DEFINED;
+	    PAGE_PTR m_page_p = nullptr;
+	    page_ptr_watcher_uptr_type m_watcher_p;
+	    int m_ref_count = -1;
+	  };
+
+	  /*
+	   * Implements a RAII-like reference counted functionality to bokkeep page pointers for
+	   * a sequence of [possibly] nested atomic replication sub-sequences.
+	   * A page can be needed by multiple levels of a nested atomic replication sequence which
+	   * perfom changes on the page. Once the page is unfixed in a sequence at a certain
+	   * level, it can be:
+	   *  - either still kept fixed if a parent [sub]sequence did the fixing and still
+	   *    needs the page
+	   *  - or unfixed if there is no parent [sub]sequence which needs the page anymore (aka:
+	   *    the [sub]sequence which just requested the fix is the outer-most one that needed
+	   *    the page in the current overall sequence of possibly nested [sub]sequences
+	   */
+	  struct page_ptr_bookkeeping
 	  {
-	    public:
-	      atomic_log_entry () = delete;
-	      atomic_log_entry (log_lsa lsa, VPID vpid, LOG_RCVINDEX rcvindex);
-
-	      atomic_log_entry (const atomic_log_entry &) = delete;
-	      atomic_log_entry (atomic_log_entry &&that);
-
-	      ~atomic_log_entry ();
-
-	      atomic_log_entry &operator= (const atomic_log_entry &) = delete;
-	      atomic_log_entry &operator= (atomic_log_entry &&) = delete;
-
-	      void apply_log_redo (THREAD_ENTRY *thread_p, log_rv_redo_context &redo_context);
-	      template <typename T>
-	      void apply_log_by_type (THREAD_ENTRY *thread_p, log_rv_redo_context &redo_context, LOG_RECTYPE rectype);
-	      int fix_page (THREAD_ENTRY *thread_p);
-	      void unfix_page (THREAD_ENTRY *thread_p);
-	      PAGE_PTR get_page_ptr ();
-	      void set_page_ptr (const PAGE_PTR &ptr);
-	      LOG_LSA get_lsa () const;
-
-	      const VPID m_vpid;
-	    private:
-	      const log_lsa m_record_lsa;
-	      const LOG_RCVINDEX m_record_index;
-	      PAGE_PTR m_page_ptr;
-
-	      std::unique_ptr<PGBUF_WATCHER> m_watcher_p;
+	    page_ptr_bookkeeping () = default;
+	    ~page_ptr_bookkeeping ();
+
+	    page_ptr_bookkeeping (const page_ptr_bookkeeping &) = delete;
+	    page_ptr_bookkeeping (page_ptr_bookkeeping &&) = delete;
+
+	    page_ptr_bookkeeping &operator= (const page_ptr_bookkeeping &) = delete;
+	    page_ptr_bookkeeping &operator= (page_ptr_bookkeeping &&) = delete;
+
+	    int fix_page (THREAD_ENTRY *thread_p, VPID vpid, LOG_RCVINDEX rcv_index, PAGE_PTR &page_ptr_out);
+	    int unfix_page (THREAD_ENTRY *thread_p, VPID vpid);
+
+	    using page_ptr_info_map_type = std::map<VPID, page_ptr_info>;
```

Comment:
Out of curiosity, why don't you use unordered_map though it doesn't have to be ordered?

Reply:
When I learned c++ there was no hash-map in STL. So, out of habit, I use map.
In this scenario, probably that it does not matter since the maps are small and the difference in performance between map and hash map is small (the extra search perfomance of hash-map versus map is compensated by the hash function).